File: backend/go/internal/protocol/control.go
package protocol

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"math"
	"regexp"
)

const (
	MaxControlFrame = 65536
	MaxPeerList     = 512
	MaxCoordinate   = 3.0e7
)

// Control error codes (spec section 5.1).
const (
	CodeIncompatibleProtocol  = "incompatible_protocol"
	CodeBadMessage            = "bad_message"
	CodeUnknownMessage        = "unknown_message"
	CodeAuthRequired          = "auth_required"
	CodeAuthFailed            = "auth_failed"
	CodeUnsupportedAuthMethod = "unsupported_auth_method"
	CodeBanned                = "banned"
	CodeRateLimited           = "rate_limited"
	CodeSessionExpired        = "session_expired"
	CodeSessionReplaced       = "session_replaced"
	CodeStaleEpoch            = "stale_epoch"
	CodeServerFull            = "server_full"
	CodeInternal              = "internal"
)

var (
	reUUID     = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)
	reUsername = regexp.MustCompile(`^[A-Za-z0-9_]{1,16}$`)
	reNetwork  = regexp.MustCompile(`^[0-9a-z:._-]{1,64}$`)
	reWorld    = regexp.MustCompile(`^[0-9a-z:._/-]{1,128}$`)
)

// ControlError is returned by ParseControl.
type ControlError struct {
	Code    string
	Message string
}

func (e *ControlError) Error() string { return e.Code + ": " + e.Message }

func bad(msg string) *ControlError { return &ControlError{Code: CodeBadMessage, Message: msg} }

type VersionInfo struct {
	Major *uint32 `json:"major"`
	Minor *uint32 `json:"minor"`
}

type ClientInfo struct {
	Name      string `json:"name"`
	Version   string `json:"version"`
	Minecraft string `json:"minecraft"`
	Loader    string `json:"loader"`
}

type Hello struct {
	Protocol     *VersionInfo `json:"protocol"`
	Client       *ClientInfo  `json:"client"`
	Capabilities []string     `json:"capabilities"`
}

type Auth struct {
	Method   *string `json:"method"`
	Username *string `json:"username"`
	UUID     *string `json:"uuid"`
}

type Resume struct {
	ResumeToken *string `json:"resume_token"`
}

type Scope struct {
	Epoch       *uint32 `json:"epoch"`
	InWorld     *bool   `json:"in_world"`
	NetworkID   *string `json:"network_id"`
	WorldID     *string `json:"world_id"`
	Attestation *string `json:"attestation"`
}

type Pos struct {
	Epoch *uint32  `json:"epoch"`
	X     *float64 `json:"x"`
	Y     *float64 `json:"y"`
	Z     *float64 `json:"z"`
}

type Peers struct {
	Epoch *uint32  `json:"epoch"`
	Rev   *uint32  `json:"rev"`
	Full  []string `json:"full"`
}

type PeersDelta struct {
	Epoch  *uint32  `json:"epoch"`
	Base   *uint32  `json:"base"`
	Rev    *uint32  `json:"rev"`
	Add    []string `json:"add"`
	Remove []string `json:"remove"`
}

type State struct {
	Muted    *bool `json:"muted"`
	Deafened *bool `json:"deafened"`
}

type Ping struct {
	Nonce *uint64 `json:"nonce"`
}

type Bye struct{}

// ParseControl parses and validates one client -> backend control frame.
// It returns one of *Hello, *Auth, *Resume, *Scope, *Pos, *Peers, *PeersDelta,
// *State, *Ping, *Bye.
func ParseControl(frame []byte) (any, *ControlError) {
	if len(frame) > MaxControlFrame {
		return nil, bad("frame too large")
	}
	trimmed := bytes.TrimLeft(frame, " \t\r\n")
	if len(trimmed) == 0 || trimmed[0] != '{' {
		return nil, bad("frame must be a JSON object")
	}
	var head struct {
		Type *string `json:"type"`
	}
	if err := json.Unmarshal(frame, &head); err != nil {
		return nil, bad("invalid JSON")
	}
	if head.Type == nil {
		return nil, bad("missing type")
	}
	var msg any
	switch *head.Type {
	case "hello":
		msg = &Hello{}
	case "auth":
		msg = &Auth{}
	case "resume":
		msg = &Resume{}
	case "scope":
		msg = &Scope{}
	case "pos":
		msg = &Pos{}
	case "peers":
		msg = &Peers{}
	case "peers_delta":
		msg = &PeersDelta{}
	case "state":
		msg = &State{}
	case "ping":
		msg = &Ping{}
	case "bye":
		return &Bye{}, nil
	default:
		return nil, &ControlError{Code: CodeUnknownMessage, Message: "unknown message type"}
	}
	if err := json.Unmarshal(frame, msg); err != nil {
		return nil, bad("field has wrong type or range")
	}
	if cerr := validate(msg); cerr != nil {
		return nil, cerr
	}
	return msg, nil
}

func shortString(s string, max int) bool { return len(s) <= max }

func validate(m any) *ControlError {
	switch v := m.(type) {
	case *Hello:
		if v.Protocol == nil || v.Protocol.Major == nil || v.Protocol.Minor == nil {
			return bad("missing protocol version")
		}
		if v.Client == nil {
			return bad("missing client info")
		}
		c := v.Client
		if !shortString(c.Name, 64) || !shortString(c.Version, 64) || !shortString(c.Minecraft, 64) || !shortString(c.Loader, 64) {
			return bad("client info too long")
		}
		if len(v.Capabilities) > 64 {
			return bad("too many capabilities")
		}
		for _, cp := range v.Capabilities {
			if !shortString(cp, 64) {
				return bad("capability too long")
			}
		}
		if *v.Protocol.Major != Major {
			return &ControlError{Code: CodeIncompatibleProtocol, Message: "unsupported protocol major version"}
		}
	case *Auth:
		if v.Method == nil || !shortString(*v.Method, 32) {
			return bad("missing auth method")
		}
		if v.Username == nil || !reUsername.MatchString(*v.Username) {
			return bad("invalid username")
		}
		if v.UUID != nil && !reUUID.MatchString(*v.UUID) {
			return bad("invalid uuid")
		}
		if *v.Method == "offline" && v.UUID == nil {
			return bad("offline auth requires uuid")
		}
	case *Resume:
		if v.ResumeToken == nil || len(*v.ResumeToken) == 0 || len(*v.ResumeToken) > 4096 {
			return bad("invalid resume token")
		}
	case *Scope:
		if v.Epoch == nil || v.InWorld == nil {
			return bad("missing epoch or in_world")
		}
		if *v.InWorld {
			if v.NetworkID == nil || !reNetwork.MatchString(*v.NetworkID) {
				return bad("invalid network_id")
			}
			if v.WorldID == nil || !reWorld.MatchString(*v.WorldID) {
				return bad("invalid world_id")
			}
		}
		if v.Attestation != nil && len(*v.Attestation) > 2048 {
			return bad("attestation too long")
		}
	case *Pos:
		if v.Epoch == nil || v.X == nil || v.Y == nil || v.Z == nil {
			return bad("missing position field")
		}
		for _, c := range []float64{*v.X, *v.Y, *v.Z} {
			if math.IsNaN(c) || math.IsInf(c, 0) || math.Abs(c) > MaxCoordinate {
				return bad("coordinate out of range")
			}
		}
	case *Peers:
		if v.Epoch == nil || v.Rev == nil || v.Full == nil {
			return bad("missing peers field")
		}
		if e := validUUIDList(v.Full); e != nil {
			return e
		}
	case *PeersDelta:
		if v.Epoch == nil || v.Base == nil || v.Rev == nil {
			return bad("missing peers_delta field")
		}
		if e := validUUIDList(v.Add); e != nil {
			return e
		}
		if e := validUUIDList(v.Remove); e != nil {
			return e
		}
	case *State:
		if v.Muted == nil || v.Deafened == nil {
			return bad("missing state field")
		}
	case *Ping:
		if v.Nonce == nil {
			return bad("missing nonce")
		}
	}
	return nil
}

func validUUIDList(l []string) *ControlError {
	if len(l) > MaxPeerList {
		return bad("peer list too long")
	}
	for _, u := range l {
		if !reUUID.MatchString(u) {
			return bad("invalid uuid in peer list")
		}
	}
	return nil
}

// ValidUUID reports whether s is a canonical lowercase UUID string.
func ValidUUID(s string) bool { return reUUID.MatchString(s) }

// UUIDBytes converts a canonical UUID string to 16 bytes.
func UUIDBytes(s string) ([16]byte, bool) {
	var out [16]byte
	if !reUUID.MatchString(s) {
		return out, false
	}
	h := s[0:8] + s[9:13] + s[14:18] + s[19:23] + s[24:36]
	if _, err := hex.Decode(out[:], []byte(h)); err != nil {
		return out, false
	}
	return out, true
}

// UUIDString formats 16 bytes as a canonical UUID string.
func UUIDString(b [16]byte) string {
	h := hex.EncodeToString(b[:])
	return h[0:8] + "-" + h[8:12] + "-" + h[12:16] + "-" + h[16:20] + "-" + h[20:32]
}

// --- backend -> client messages ---

type ErrorMsg struct {
	Type    string `json:"type"`
	Code    string `json:"code"`
	Message string `json:"message"`
	Fatal   bool   `json:"fatal"`
}

type ServerInfo struct {
	Name           string `json:"name"`
	Version        string `json:"version"`
	Implementation string `json:"implementation"`
}

type AuthInfo struct {
	Methods   []string `json:"methods"`
	Challenge string   `json:"challenge"`
}

type Version struct {
	Major int `json:"major"`
	Minor int `json:"minor"`
}

type HelloOK struct {
	Type         string     `json:"type"`
	Protocol     Version    `json:"protocol"`
	Server       ServerInfo `json:"server"`
	Capabilities []string   `json:"capabilities"`
	Auth         AuthInfo   `json:"auth"`
}

type VoiceInfo struct {
	Host         string `json:"host"`
	Port         int    `json:"port"`
	ConnectionID string `json:"connection_id"`
	KeyID        int    `json:"key_id"`
	Key          string `json:"key"`
	KeyExpiresIn int    `json:"key_expires_in"`
}

type SessionConfig struct {
	NormalRange       float64 `json:"normal_range"`
	WhisperRange      float64 `json:"whisper_range"`
	MaxRange          float64 `json:"max_range"`
	Codec             string  `json:"codec"`
	SampleRate        int     `json:"sample_rate"`
	FrameMs           int     `json:"frame_ms"`
	HeartbeatInterval int     `json:"heartbeat_interval"`
	PositionHzMax     int     `json:"position_hz_max"`
}

type SessionMsg struct {
	Type        string        `json:"type"`
	SessionID   string        `json:"session_id"`
	PlayerUUID  string        `json:"player_uuid"`
	Username    string        `json:"username"`
	ResumeToken string        `json:"resume_token"`
	Voice       VoiceInfo     `json:"voice"`
	Config      SessionConfig `json:"config"`
}

type KeyMsg struct {
	Type         string `json:"type"`
	KeyID        int    `json:"key_id"`
	Key          string `json:"key"`
	KeyExpiresIn int    `json:"key_expires_in"`
}

type PresenceMsg struct {
	Type   string   `json:"type"`
	Epoch  uint32   `json:"epoch"`
	Add    []string `json:"add"`
	Remove []string `json:"remove"`
}

type PeersResync struct {
	Type  string `json:"type"`
	Epoch uint32 `json:"epoch"`
}

type Pong struct {
	Type       string `json:"type"`
	Nonce      uint64 `json:"nonce"`
	ServerTime int64  `json:"server_time"`
}

type Simple struct {
	Type string `json:"type"`
}

// Capabilities advertised by both backend implementations.
var ServerCapabilities = []string{"opus", "whisper", "peers_delta", "presence", "key_rotation", "scope_attestation"}

File: backend/rust/src/protocol/control.rs
//! Control channel messages (spec section 6).

use serde::{Deserialize, Serialize};

pub const MAX_CONTROL_FRAME: usize = 65536;
pub const MAX_PEER_LIST: usize = 512;
pub const MAX_COORDINATE: f64 = 3.0e7;

pub mod code {
    pub const INCOMPATIBLE_PROTOCOL: &str = "incompatible_protocol";
    pub const BAD_MESSAGE: &str = "bad_message";
    pub const UNKNOWN_MESSAGE: &str = "unknown_message";
    pub const AUTH_REQUIRED: &str = "auth_required";
    pub const AUTH_FAILED: &str = "auth_failed";
    pub const UNSUPPORTED_AUTH_METHOD: &str = "unsupported_auth_method";
    pub const BANNED: &str = "banned";
    pub const RATE_LIMITED: &str = "rate_limited";
    pub const SESSION_EXPIRED: &str = "session_expired";
    pub const SESSION_REPLACED: &str = "session_replaced";
    pub const STALE_EPOCH: &str = "stale_epoch";
    pub const SERVER_FULL: &str = "server_full";
    pub const INTERNAL: &str = "internal";
}

pub const SERVER_CAPABILITIES: &[&str] = &["opus", "whisper", "peers_delta", "presence", "key_rotation", "scope_attestation"];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ControlError {
    pub code: &'static str,
    pub message: &'static str,
}

fn bad(message: &'static str) -> ControlError {
    ControlError {
        code: code::BAD_MESSAGE,
        message,
    }
}

#[derive(Debug, Deserialize)]
pub struct VersionInfo {
    pub major: Option<u32>,
    pub minor: Option<u32>,
}

#[derive(Debug, Deserialize, Default)]
#[serde(default)]
pub struct ClientInfo {
    pub name: String,
    pub version: String,
    pub minecraft: String,
    pub loader: String,
}

#[derive(Debug, Deserialize)]
pub struct Hello {
    pub protocol: Option<VersionInfo>,
    pub client: Option<ClientInfo>,
    #[serde(default)]
    pub capabilities: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
pub struct Auth {
    pub method: Option<String>,
    pub username: Option<String>,
    pub uuid: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct Resume {
    pub resume_token: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct Scope {
    pub epoch: Option<u32>,
    pub in_world: Option<bool>,
    pub network_id: Option<String>,
    pub world_id: Option<String>,
    pub attestation: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct Pos {
    pub epoch: Option<u32>,
    pub x: Option<f64>,
    pub y: Option<f64>,
    pub z: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub struct Peers {
    pub epoch: Option<u32>,
    pub rev: Option<u32>,
    pub full: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
pub struct PeersDelta {
    pub epoch: Option<u32>,
    pub base: Option<u32>,
    pub rev: Option<u32>,
    pub add: Option<Vec<String>>,
    pub remove: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
pub struct State {
    pub muted: Option<bool>,
    pub deafened: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct Ping {
    pub nonce: Option<u64>,
}

/// A validated client -> backend control message.
#[derive(Debug)]
pub enum ClientMsg {
    Hello(Hello),
    Auth(Auth),
    Resume(Resume),
    Scope(Scope),
    Pos(Pos),
    Peers(Peers),
    PeersDelta(PeersDelta),
    State(State),
    Ping(Ping),
    Bye,
}

#[derive(Deserialize)]
struct Head {
    #[serde(rename = "type")]
    t: Option<String>,
}

fn parse_as<'a, T: Deserialize<'a>>(frame: &'a [u8]) -> Result<T, ControlError> {
    serde_json::from_slice(frame).map_err(|_| bad("field has wrong type or range"))
}

/// Parse and validate one client control frame.
pub fn parse_control(frame: &[u8]) -> Result<ClientMsg, ControlError> {
    if frame.len() > MAX_CONTROL_FRAME {
        return Err(bad("frame too large"));
    }
    match frame.iter().find(|b| !b" \t\r\n".contains(b)) {
        Some(b'{') => {}
        _ => return Err(bad("frame must be a JSON object")),
    }
    let head: Head = serde_json::from_slice(frame).map_err(|_| bad("invalid JSON"))?;
    let t = head.t.ok_or_else(|| bad("missing type"))?;
    let msg = match t.as_str() {
        "hello" => ClientMsg::Hello(parse_as(frame)?),
        "auth" => ClientMsg::Auth(parse_as(frame)?),
        "resume" => ClientMsg::Resume(parse_as(frame)?),
        "scope" => ClientMsg::Scope(parse_as(frame)?),
        "pos" => ClientMsg::Pos(parse_as(frame)?),
        "peers" => ClientMsg::Peers(parse_as(frame)?),
        "peers_delta" => ClientMsg::PeersDelta(parse_as(frame)?),
        "state" => ClientMsg::State(parse_as(frame)?),
        "ping" => ClientMsg::Ping(parse_as(frame)?),
        "bye" => return Ok(ClientMsg::Bye),
        _ => {
            return Err(ControlError {
                code: code::UNKNOWN_MESSAGE,
                message: "unknown message type",
            })
        }
    };
    validate(&msg)?;
    Ok(msg)
}

pub fn is_uuid(s: &str) -> bool {
    let b = s.as_bytes();
    b.len() == 36
        && b.iter().enumerate().all(|(i, &c)| match i {
            8 | 13 | 18 | 23 => c == b'-',
            _ => c.is_ascii_digit() || (b'a'..=b'f').contains(&c),
        })
}

fn is_username(s: &str) -> bool {
    (1..=16).contains(&s.len()) && s.bytes().all(|c| c.is_ascii_alphanumeric() || c == b'_')
}

fn is_network(s: &str) -> bool {
    (1..=64).contains(&s.len())
        && s.bytes()
            .all(|c| c.is_ascii_digit() || c.is_ascii_lowercase() || b":._-".contains(&c))
}

fn is_world(s: &str) -> bool {
    (1..=128).contains(&s.len())
        && s.bytes()
            .all(|c| c.is_ascii_digit() || c.is_ascii_lowercase() || b":._/-".contains(&c))
}

fn uuid_list(l: &[String]) -> Result<(), ControlError> {
    if l.len() > MAX_PEER_LIST {
        return Err(bad("peer list too long"));
    }
    if l.iter().all(|u| is_uuid(u)) {
        Ok(())
    } else {
        Err(bad("invalid uuid in peer list"))
    }
}

fn validate(m: &ClientMsg) -> Result<(), ControlError> {
    match m {
        ClientMsg::Hello(h) => {
            let p = h.protocol.as_ref().ok_or_else(|| bad("missing protocol version"))?;
            let (Some(major), Some(_)) = (p.major, p.minor) else {
                return Err(bad("missing protocol version"));
            };
            let c = h.client.as_ref().ok_or_else(|| bad("missing client info"))?;
            if [&c.name, &c.version, &c.minecraft, &c.loader].iter().any(|s| s.len() > 64) {
                return Err(bad("client info too long"));
            }
            if let Some(caps) = &h.capabilities {
                if caps.len() > 64 || caps.iter().any(|c| c.len() > 64) {
                    return Err(bad("too many capabilities"));
                }
            }
            if major != u32::from(super::udp::MAJOR) {
                return Err(ControlError {
                    code: code::INCOMPATIBLE_PROTOCOL,
                    message: "unsupported protocol major version",
                });
            }
        }
        ClientMsg::Auth(a) => {
            let method = a
                .method
                .as_deref()
                .filter(|m| m.len() <= 32)
                .ok_or_else(|| bad("missing auth method"))?;
            if !a.username.as_deref().is_some_and(is_username) {
                return Err(bad("invalid username"));
            }
            if let Some(u) = &a.uuid {
                if !is_uuid(u) {
                    return Err(bad("invalid uuid"));
                }
            }
            if method == "offline" && a.uuid.is_none() {
                return Err(bad("offline auth requires uuid"));
            }
        }
        ClientMsg::Resume(r) => {
            if !r.resume_token.as_deref().is_some_and(|t| !t.is_empty() && t.len() <= 4096) {
                return Err(bad("invalid resume token"));
            }
        }
        ClientMsg::Scope(s) => {
            let (Some(_), Some(in_world)) = (s.epoch, s.in_world) else {
                return Err(bad("missing epoch or in_world"));
            };
            if in_world {
                if !s.network_id.as_deref().is_some_and(is_network) {
                    return Err(bad("invalid network_id"));
                }
                if !s.world_id.as_deref().is_some_and(is_world) {
                    return Err(bad("invalid world_id"));
                }
            }
            if s.attestation.as_ref().is_some_and(|a| a.len() > 2048) {
                return Err(bad("attestation too long"));
            }
        }
        ClientMsg::Pos(p) => {
            let (Some(_), Some(x), Some(y), Some(z)) = (p.epoch, p.x, p.y, p.z) else {
                return Err(bad("missing position field"));
            };
            if [x, y, z].iter().any(|c| !c.is_finite() || c.abs() > MAX_COORDINATE) {
                return Err(bad("coordinate out of range"));
            }
        }
        ClientMsg::Peers(p) => {
            let (Some(_), Some(_), Some(full)) = (p.epoch, p.rev, &p.full) else {
                return Err(bad("missing peers field"));
            };
            uuid_list(full)?;
        }
        ClientMsg::PeersDelta(d) => {
            if d.epoch.is_none() || d.base.is_none() || d.rev.is_none() {
                return Err(bad("missing peers_delta field"));
            }
            uuid_list(d.add.as_deref().unwrap_or(&[]))?;
            uuid_list(d.remove.as_deref().unwrap_or(&[]))?;
        }
        ClientMsg::State(s) => {
            if s.muted.is_none() || s.deafened.is_none() {
                return Err(bad("missing state field"));
            }
        }
        ClientMsg::Ping(p) => {
            if p.nonce.is_none() {
                return Err(bad("missing nonce"));
            }
        }
        ClientMsg::Bye => {}
    }
    Ok(())
}

pub fn uuid_bytes(s: &str) -> Option<[u8; 16]> {
    if !is_uuid(s) {
        return None;
    }
    let h: String = s.chars().filter(|c| *c != '-').collect();
    let mut out = [0u8; 16];
    hex::decode_to_slice(h, &mut out).ok()?;
    Some(out)
}

pub fn uuid_string(b: &[u8; 16]) -> String {
    let h = hex::encode(b);
    format!("{}-{}-{}-{}-{}", &h[0..8], &h[8..12], &h[12..16], &h[16..20], &h[20..32])
}

// --- backend -> client messages ---

#[derive(Serialize)]
pub struct ErrorMsg<'a> {
    #[serde(rename = "type")]
    pub t: &'static str,
    pub code: &'a str,
    pub message: &'a str,
    pub fatal: bool,
}

pub fn error_frame(code: &str, message: &str, fatal: bool) -> String {
    serde_json::to_string(&ErrorMsg {
        t: "error",
        code,
        message,
        fatal,
    })
    .unwrap()
}

#[derive(Serialize)]
pub struct Version {
    pub major: u32,
    pub minor: u32,
}

#[derive(Serialize)]
pub struct ServerInfo {
    pub name: &'static str,
    pub version: &'static str,
    pub implementation: &'static str,
}

#[derive(Serialize)]
pub struct AuthInfo {
    pub methods: Vec<String>,
    pub challenge: String,
}

#[derive(Serialize)]
pub struct HelloOk {
    #[serde(rename = "type")]
    pub t: &'static str,
    pub protocol: Version,
    pub server: ServerInfo,
    pub capabilities: &'static [&'static str],
    pub auth: AuthInfo,
}

#[derive(Serialize)]
pub struct VoiceInfo {
    pub host: String,
    pub port: u16,
    pub connection_id: String,
    pub key_id: u8,
    pub key: String,
    pub key_expires_in: u64,
}

#[derive(Serialize)]
pub struct SessionConfig {
    pub normal_range: f64,
    pub whisper_range: f64,
    pub max_range: f64,
    pub codec: &'static str,
    pub sample_rate: u32,
    pub frame_ms: u32,
    pub heartbeat_interval: u32,
    pub position_hz_max: u32,
}

#[derive(Serialize)]
pub struct SessionMsg {
    #[serde(rename = "type")]
    pub t: &'static str,
    pub session_id: String,
    pub player_uuid: String,
    pub username: String,
    pub resume_token: String,
    pub voice: VoiceInfo,
    pub config: SessionConfig,
}

#[derive(Serialize)]
pub struct KeyMsg {
    #[serde(rename = "type")]
    pub t: &'static str,
    pub key_id: u8,
    pub key: String,
    pub key_expires_in: u64,
}

#[derive(Serialize)]
pub struct PresenceMsg {
    #[serde(rename = "type")]
    pub t: &'static str,
    pub epoch: u32,
    pub add: Vec<String>,
    pub remove: Vec<String>,
}

#[derive(Serialize)]
pub struct EpochMsg {
    #[serde(rename = "type")]
    pub t: &'static str,
    pub epoch: u32,
}

#[derive(Serialize)]
pub struct PongMsg {
    #[serde(rename = "type")]
    pub t: &'static str,
    pub nonce: u64,
    pub server_time: i64,
}
